cudaPackages.nccl: refactor to fix #220340 and #221895

ConnorBaker opened the pull request:
###### Description of changes

- moves to `autoAddOpenGLRunpathHook` instead of manual invocation of `addOpenGLRunpath`
  - closes #220340.
- corrects bug where `nccl` was only built with default version of `cudaPackages` by changing the `nccl` derivation to accept the actual CUDA packages it requires as arguments
  - closes #221895
  - fixes a bug where NCCL would not build with CUDA 12 or newer without overriding the default version of `cudaPackages` in Nixpkgs
- fixes paths given to the environment for `cuda_cudart` to prevent breakage when https://github.com/NixOS/nixpkgs/pull/240498 is merged
- fixes a bug where NCCL would not build with CUDA 12.0 due to a comparison with a version string of the wrong precision
- ~~removed unnecessary `which` entry in `nativeBuildInputs`~~
  - NOTE: While it seems like `which` is unused, it's used to locate NVCC and extract the CUDA version. This is in turn used for conditional compilation.

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

ConnorBaker commented:
Result of `nixpkgs-review pr 217619` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python310Packages.tensorflowWithCuda</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>cudaPackages.nccl</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
  </ul>
</details>

ConnorBaker commented:
Result of `nixpkgs-review pr 217619` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package failed to build:</summary>
  <ul>
    <li>python310Packages.tensorflowWithCuda</li>
  </ul>
</details>
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>cudaPackages.nccl</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.jaxlibWithCuda</li>
  </ul>
</details>

ConnorBaker commented:
Waiting for https://github.com/NixOS/nixpkgs/pull/220402 to be merged prior to proceeding.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/tweag-nix-dev-update-45/26397/1

ConnorBaker commented:
Result of `nixpkgs-review pr 217619 --extra-nixpkgs-config '{ allowUnfree = true; cudaSupport = true; cudaForwardCompat = false; cudaCapabilities = [ "8.6" ]; }'` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>16 packages built:</summary>
  <ul>
    <li>cudaPackages.nccl</li>
    <li>cudaPackages.nccl.dev</li>
    <li>python310Packages.cupy</li>
    <li>python310Packages.cupy.dist</li>
    <li>python310Packages.jaxlibWithCuda</li>
    <li>python310Packages.jaxlibWithCuda.dist</li>
    <li>python310Packages.tensorflowWithCuda</li>
    <li>python310Packages.tensorflowWithCuda.dist</li>
    <li>python310Packages.torchWithCuda</li>
    <li>python310Packages.torchWithCuda.dev</li>
    <li>python310Packages.torchWithCuda.dist</li>
    <li>python310Packages.torchWithCuda.lib</li>
    <li>python311Packages.cupy</li>
    <li>python311Packages.cupy.dist</li>
    <li>python311Packages.jaxlibWithCuda</li>
    <li>python311Packages.jaxlibWithCuda.dist</li>
  </ul>
</details>

ConnorBaker commented:
CC @NixOS/cuda-maintainers

samuela reviewed (commented):
I'm not familiar with what the runtime dependencies of nccl are, but these changes look reasonable to me. Perhaps @mdaiter or @orivej could comment?

As we move towards more of a "python-packages.nix" model, I think it might be worth collecting all `cudaPackages.*` packages into a single directory like python does. We have a bunch of random stuff spread out all over the codebase for no good reason except that that's just how things evolved over time. But we don't need to keep living like this. NCCL isn't really math software anyhow.

ConnorBaker commented:
> As we move towards more of a "python-packages.nix" model, I think it might be worth collecting all `cudaPackages.*` packages into a single directory like python does. We have a bunch of random stuff spread out all over the codebase for no good reason except that that's just how things evolved over time. But we don't need to keep living like this. NCCL isn't really math software anyhow.

I like that idea! It would certainly make it easier to keep track of different components.

How do you envision stuff like `torch` or `magma` working? They're packages which can rely on CUDA, but (at least `torch`) perhaps doesn't belong in the CUDA packages directory.

As a separate point, in the same way that we can get `torch` with different versions of python through `python3*Packages`, should we likewise have `cudaPackages_*.torch` available? If yes, how would we handle different versions of Python? If we were to want to match Anaconda, the user should be able to specify the version of CUDA and Python for `torch`. Of course this can be done by using overlays and changing what `cudaPackages` and `python3Packages` points to, but that's not necessarily user friendly if it's hard to discover/arrive at. (Although maybe this is tangential -- not enough coffee yet. If it is, let me know and I can throw it in a separate issue to track/discuss further!)

samuela commented:
> How do you envision stuff like torch or magma working? They're packages which can rely on CUDA, but (at least torch) perhaps doesn't belong in the CUDA packages directory.

Good question! The line is a little fuzzy but I envision that the general policy would be that `cudaPackages.*` would be only for packages that are part of CUDA toolkit or related packages from NVIDIA, and all consumers of those packages (torch, magma, etc) would continue to live externally. This would more or less match the current Python package situation AFAIU it -- eg C++ packages that rely on python packages are still maintained outside of `pkgs/development/python-modules`.

> As a separate point, in the same way that we can get torch with different versions of python through python3*Packages, should we likewise have cudaPackages_*.torch available? If yes, how would we handle different versions of Python? If we were to want to match Anaconda, the user should be able to specify the version of CUDA and Python for torch. Of course this can be done by using overlays and changing what cudaPackages and python3Packages points to, but that's not necessarily user friendly if it's hard to discover/arrive at. (Although maybe this is tangential -- not enough coffee yet. If it is, let me know and I can throw it in a separate issue to track/discuss further!)

I believe this may indeed be a tangential point :P Actually, I think we should focus on supporting fewer version combinations, not more. Having to maintain support for all combinations increases testing and maintenance burden. Users who are interested in using older versions can always use custom overlays or pin nixpkgs to previous commits.

orivej commented:
> I'm not familiar with what the runtime dependencies of nccl are, but these changes look reasonable to me. Perhaps @mdaiter or @orivej could comment?

nccl does not have any runtime dependencies.

The changes look good to me (although I'm not familiar with the concept of backendStdenv and don't understand why build dependencies have to be moved to nativeBuildInputs).

SomeoneSerge commented on pkgs/development/libraries/science/math/nccl/default.nix:
```diff
@@ -1,46 +1,40 @@
-    addOpenGLRunpath
+    autoAddOpenGLRunpathHook
+    cuda_cccl #include <nv/target>
+    cuda_cudart #include cuda_runtime.h
```
I assume this is what @orivej  refers to: cudart and cccl are headers and shared libs for the target platform, we should try and have them in `buildInputs`

ConnorBaker replied:
Ah I thought `nativeBuildInputs` was for dependencies required at build-time, and `buildInputs` for those required at runtime? If that's not the case, let me know and I'll fix it!

SomeoneSerge replied:
Yea, it's a bit more complicated than that and we're being pretty lousy with this only because we don't cross-compile cuda-related packages, we build on the same platform we build _for_

The `nativeBuildInputs` is for the binaries (usually the toolchain) you'd have to execute on your build machine.
The `buildInputs` are for binaries that are to be executed on the target machine, the one you deploy your build artifact at. E.g. if your build machine is x84_64, but you cross-compile for aarch64, you'll link an aarch64 version of `libcudart.so`

In practice, I suspect most of our cuda stuff is not cross-compatible at the moment, not least because of ambiguous string interpolations. But I also just don't know how cross in nixpkgs actually works

SomeoneSerge commented:
Thanks @orivej, `backendStdenv` is just a hacky way to refer to a nvcc-compatible gccStdenv. We had some libstdc++ mismatch issues

ConnorBaker commented:
Rebased, moved `cuda_cccl` and `cuda_cudart` to `buildInputs` from `nativeBuildInputs`, and changed license from `bsd3` to `unfreeRedistributable`. Running `nixpkgs-review` now!

SomeoneSerge commented on pkgs/development/libraries/science/math/nccl/default.nix:
```diff
@@ -53,25 +50,18 @@ backendStdenv.mkDerivation rec {
     description = "Multi-GPU and multi-node collective communication primitives for NVIDIA GPUs";
     homepage = "https://developer.nvidia.com/nccl";
-    license = licenses.bsd3;
+    # Though the NCCL has a BSD3 license, it uses CUDA which is not free.
```
This is an interesting one. In case of nccl we build from source and then link against the unfree cuda stuff. Nccl nix expression already won't evaluate, because it consumes cudaPackages. I think this is very different e.g. from torch-bin, where cuda bits are bundled in and there's nothing to prevent evaluation

orivej replied:
Yeah, I think the license should be kept as BSD3, because this is what NCCL is licensed by.

ConnorBaker replied:
Oh! Okay, that makes sense — will fix today.

ConnorBaker replied:
Fixed in newest commit

orivej replied:
Currently it still says `license = licenses.unfreeRedistributable;`

ConnorBaker replied:
Ugh. I forgot to do `git add`.

ConnorBaker replied:
Should definitely be correct now!

ConnorBaker commented:
I let it run overnight on my i9 13900K and it still didn't finish -- hopefully this time around it'll be faster :x

samuela commented on pkgs/development/libraries/science/math/nccl/default.nix:
```diff
@@ -1,46 +1,43 @@
-    cuda_cudart
-  ] ++ lib.optionals (lib.versionAtLeast cudaVersion "12.0.0") [
-    cuda_cccl
+    cuda_cccl #include <nv/target>
```
previously `cuda_cccl` was restricted to cudaVersion>=12. is there a reason to add it to builds for all versions?

ConnorBaker replied:
When I had tested it I needed it on all versions of CUDA iirc. I can test again later but I remember getting errors until I did that.

ConnorBaker replied:
@connorbaker reminder to test when I get a chance

ConnorBaker replied:
It was, in fact, required only for CUDA 12.0 and later.

The error I was running into was actually a bug due to the comparison with a version number with three components -- `versionAtLeast "12.0" "12.0.0"` evaluates to `false` so I was seeing build failures with `cudaPackages_12_0`.

SomeoneSerge commented on pkgs/development/libraries/science/math/nccl/default.nix:
```diff
@@ -1,46 +1,43 @@
-    addOpenGLRunpath
+    autoAddOpenGLRunpathHook
     cuda_nvcc
+    which
```
What's it use which for again? Can we control it w/o which?

samuela commented:
> I don't remember where we stalled, but this LGTM. Nixpkgs-review and merge?

I believe @ConnorBaker left a TODO item for himself [here](https://github.com/NixOS/nixpkgs/pull/217619#discussion_r1150031387), but not sure what the status of that is?

ConnorBaker committed:
nccl: use packages from cudaPackages

The previous usage of `cudaPackages` ensured that we only ever saw packages from
the default version of `cudaPackages` Nixpkgs uses.

ConnorBaker commented:
Found out this was still open when I tried to update NCCL to 2.18.3.

Let me try to pick this back up :)

ConnorBaker commented:
Rebased, updated, and added more notes about what it's fixing.

Testing with a super simple test suite built around `nccl-tests`: https://github.com/ConnorBaker/nix-cuda-test

```bash
nix build -L --override-input nixpkgs github:nixos/nixpkgs/pull/217619/head github:ConnorBaker/nix-cuda-test/3074628f4a0ece4928e032f9e2f2f1307b6ed22d#nccl-test-suite
```

Seemed to work for me, though it's just one device.

https://gist.github.com/ConnorBaker/ea3c49e23c1eaf2544fe97ae6fdd67a5
